File: solutions/python/dp-1d/last-stone-weight-ii.py
# LeetCode 1049 — Last Stone Weight II (Medium)
# Category: 1-D Dynamic Programming · Approach: Subset Sum
# Time: O(n * sum) | Space: O(sum)
# Source: https://leetcode.com/problems/last-stone-weight-ii/

def last_stone_weight_ii(stones: list[int]) -> int:
    total = sum(stones)
    t = total // 2
    dp = [False] * (t + 1)
    dp[0] = True
    for x in stones:
        for j in range(t, x - 1, -1):
            dp[j] = dp[j] or dp[j - x]
    for j in range(t, -1, -1):
        if dp[j]:
            return total - 2 * j
    return total

File: solutions/typescript/dp-1d/last-stone-weight-ii.ts
// LeetCode 1049 — Last Stone Weight II (Medium)
// Category: 1-D Dynamic Programming · Approach: Subset Sum
// Time: O(n * sum) | Space: O(sum)
// Source: https://leetcode.com/problems/last-stone-weight-ii/

function lastStoneWeightII(stones: number[]): number {
  const total = stones.reduce((a, b) => a + b, 0);
  const t = Math.floor(total / 2);
  const dp = new Array<boolean>(t + 1).fill(false);
  dp[0] = true;
  for (const x of stones) {
    for (let j = t; j >= x; j--) dp[j] = dp[j] || dp[j - x];
  }
  for (let j = t; j >= 0; j--) if (dp[j]) return total - 2 * j;
  return total;
}
